### PR TITLE
324 akka dispatcher total threads active threads

### LINKED ISSUE #323: Dispatcher metrics

### PR DESCRIPTION
**Pull request by ornel-lloyd-edano**
Closes #323 #324
___

## Summary:

add dispatcher metrics:
- static max threads config
- static min threads config
- static parallelism config
- active thread count
- total thread count

### CODE REVIEW

**Review by lgajowy — COMMENTED** [review]
Some initial comments. I will continue reviewing tomorrow.

**Comment by lgajowy** [comment]
@ornel-lloyd-edano please also update this branch and resolve the conflicts

File: agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/ActorCellConstructorAdvice.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  def constructor(
+    @Argument(0) system: akka.actor.ExtendedActorSystem,
+    @Argument(3) dispatcher: MessageDispatcher
+  ): Unit =
+    if (dispatcher.id == "akka.actor.default-dispatcher") {
+      DispatcherDecorator.setActorSystem(dispatcher.asInstanceOf[Dispatcher], system)
+
+      val akkaConfig       = system.settings.config.getConfig("akka.actor")
+      val dispatcherConfig = akkaConfig.getConfig("default-dispatcher")
+      dispatcherConfig.getString("executor") match {
```
Comment (lgajowy) [inline agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/ActorCellConstructorAdvice.scala:24]: I still wonder about these _static_ props... If we always take these from the config and they never change, does it make sense to make a Mesmer feature out of it at all? After all users can reach their configs. Unless there are scenarios when these properties can change but I'm not aware?

@ornel-lloyd-edano @worekleszczy what do you think?

Comment (lgajowy) [reply agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/ActorCellConstructorAdvice.scala:24]: Also, notice that other solutions (Kamon) do not rely on the config (at least directly). Not sure why (I would need to investigate myself).

File: agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/AkkaDispatcherAgent.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      minThreads = minThreadsAgent.isDefined,
+      maxThreads = maxThreadsAgent.isDefined,
+      parallelismFactor = parallelismFactorAgent.isDefined,
+      totalThreads = totalThreadsAgent.isDefined,
+      activeThreads = activeThreadsAgent.isDefined
+    )
+    (resultantAgent, enabled)
+  }
+
+  private def ifSupported(versions: AkkaDispatcherModule.Jars[Version])(agent: => Agent): Option[Agent] =
```
[the remark this answers is not in the document]
Comment (ornel-lloyd-edano) [reply agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/AkkaDispatcherAgent.scala]: done

File: agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/DispatcherDecorator.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package io.scalac.mesmer.agent.akka.dispatcher
+
+import akka.actor.ActorSystem
+import akka.dispatch.Dispatcher
+
+object DispatcherDecorator {
+  private val map = scala.collection.mutable.Map[Dispatcher, ActorSystem]()
```
[the remark this answers is not in the document]
Comment (ornel-lloyd-edano) [reply agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/DispatcherDecorator.scala:7]: I initially used the Virtual Fields but for some reason when I fetch the actor system I am getting a null so I ended up using just a plain mutable Map. But I will try to experiment with it again maybe something was off during setting the value using classOf

File: agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/ActorCellConstructorAdvice.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import io.scalac.mesmer.core.event.DispatcherEvent.ExecutorConfigEvent
+import io.scalac.mesmer.core.event.EventBus
+
+object ActorCellConstructorAdvice {
+
+  @OnMethodEnter
+  def constructor(
+    @Argument(0) system: akka.actor.ExtendedActorSystem,
+    @Argument(3) dispatcher: MessageDispatcher
+  ): Unit =
```
Comment (worekleszczy) [inline agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/ActorCellConstructorAdvice.scala:18]: Imo this is a NOGO - we'd generate a lot of events (as this executed in the constructor of `akka.actor.ActorCell`) for single execution context (most of the actors share the ec - there is not point to generate this event for every actor)

File: agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/DispatcherExecuteTaskAdvice.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import io.scalac.mesmer.core.event.EventBus
+
+object DispatcherExecuteTaskAdvice {
+  @OnMethodExit
+  def executeTask(
+    @This self: Any, // note: putting Dispatcher here instead of Any results to classloader exception
+    @FieldValue("executorServiceDelegate") executorService: ExecutorServiceDelegate
+  ): Unit =
+    self match {
+      case dispatcher: Dispatcher if dispatcher.id == "akka.actor.default-dispatcher" =>
```
Comment (worekleszczy) [inline agent/src/main/scala/io/scalac/mesmer/agent/akka/dispatcher/DispatcherExecuteTaskAdvice.scala:25]: We should support all of the dispatchers

**Pull request closed without merging**